core-rust/state-manager/src/store/proofs_gc.rs

```diff
@@ -0,0 +1,223 @@
+        // A more robust solution should either use DB snapshots (a development direction we are
+        // considering anyway), or a more selective lock over the ledger proofs CF alone (or even
+        // over a selected *range* of the ledger proofs CF).
+
+        // Read the GC's persisted state and initialize the run:
+        let read_progress_database = self.database.read_current();
+        let to_epoch = read_progress_database
+            .max_epoch()
+            .map(|current_epoch| current_epoch.number())
+            .and_then(|current| current.checked_sub(self.most_recent_full_resolution_epoch_count))
+            .map(Epoch::of);
+        let Some(to_epoch) = to_epoch else {
+            // Nothing to GC ever, yet.
+            return;
+        };
+        let progress_started_at: LedgerProofsGcProgress = read_progress_database
+            .get_progress()
+            .unwrap_or_else(LedgerProofsGcProgress::none);
+        drop(read_progress_database);
+
+        if progress_started_at.last_pruned_epoch >= to_epoch {
+            // Nothing to GC during this run.
+            return;
+        }
+
+        info!(
+            "Starting a GC run: pruning ledger proofs up to epoch {}; current progress: {:?}",
+            to_epoch.number(),
+            progress_started_at,
+        );
+        let mut last_pruned_state_version = progress_started_at.epoch_proof_state_version;
+
+        let mut retained_proofs = 0; // only for logging purposes
+        loop {
+            let at_state_version = last_pruned_state_version
+                .next()
+                .expect("state version overflow");
+
+            // Locate the next proof that we need to retain. We use the same method and limits as
+            // the business logic responsible for outputting proven transactions:
+            let locate_proof_database = self.database.read_current();
+            let transactions_and_proof = locate_proof_database.get_txns_and_proof(
+                at_state_version,
+                MAX_TXNS_FOR_RESPONSES_SPANNING_MORE_THAN_ONE_PROOF,
+                MAX_TXN_BYTES_FOR_A_SINGLE_RESPONSE,
+            );
+            drop(locate_proof_database);
+
+            let Some((_transactions, proof)) = transactions_and_proof else {
+                error!(
+                    "A chain of transactions-without-proof from state version {} does not fit within the limits; aborting the GC",
+                    at_state_version,
+                );
+                return;
+            };
+            let header = proof.ledger_header;
+
+            // Delete all the proofs from the beginning of transactions listing up to (but
+            // excluding) the returned proof:
+            last_pruned_state_version = header.state_version;
+            let delete_database = self.database.read_current();
+            delete_database.delete_ledger_proofs_range(at_state_version, last_pruned_state_version);
```

Comment:
Potential issue (!!) (EDIT: Not actually an issue I think, but still a point of confusion).

The Rust doc says "from inclusive, to exclusive" - but if we delete the proof `at_state_version` don't we delete the previous end-proof and break our invariant? 

Maybe I'm misunderstanding? EDIT: Okay, I see my mistake, I has assumed `at_state_version` was `last_pruned_state_version` but actually we do:
```rust
let at_state_version = last_pruned_state_version
                .next()
```

However I believe this naming is confusing - for me, a ledger state sync simulating "at state version X" would have its ledger tip at `X` and should be requesting with a `start_state_version_inclusive` of X + 1.

Could I suggest therefore renaming the `at_state_version` variable to `batch_start_state_version_inclusive` or similar? This then aligns with the naming of the corresponding parameter to `get_txns_and_proof` and `delete_ledger_proofs_range` and makes every line-by-line read correctly to me.

Reply:
Regardless, I'd like to see a test which can commit stuff, and then can run the GC then still serve ledger syncs.

_Or_ if it's easier, can commit a few epochs, then wipe the GC tracker, start again from scratch, and not hit the `"A chain of transactions-without-proof from state version {} does not fit within the limits; aborting the GC"` error.

Possibly updating the Java test might be easier.